pkg/validate/scoped/hydrate/namespace_hydrator.go

```diff
@@ -15,85 +15,148 @@
 package hydrate
 
 import (
+	"context"
+	"fmt"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
 	v1 "kpt.dev/configsync/pkg/api/configmanagement/v1"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/declared"
 	"kpt.dev/configsync/pkg/importer/analyzer/ast"
 	"kpt.dev/configsync/pkg/importer/analyzer/transform/selectors"
 	"kpt.dev/configsync/pkg/kinds"
 	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+	"kpt.dev/configsync/pkg/rootsync"
 	"kpt.dev/configsync/pkg/status"
 	"kpt.dev/configsync/pkg/validate/objects"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 // NamespaceSelectors hydrates the given Scoped objects by performing namespace
 // selection to copy objects into namespaces which match their selector. It also
 // sets a default namespace on any namespace-scoped object that does not already
 // have a namespace or namespace selector set.
-func NamespaceSelectors(objs *objects.Scoped) status.MultiError {
-	nsSelectors, errs := buildSelectorMap(objs)
-	if errs != nil {
-		return errs
-	}
-	var result []ast.FileObject
-	for _, obj := range objs.Namespace {
-		_, hasSelector := obj.GetAnnotations()[metadata.NamespaceSelectorAnnotationKey]
-		if hasSelector {
-			copies, err := makeNamespaceCopies(obj, nsSelectors)
-			if err != nil {
-				errs = status.Append(errs, err)
+func NamespaceSelectors(ctx context.Context, c client.Client) objects.ScopedVisitor {
+	return func(objs *objects.Scoped) status.MultiError {
+		nsSelectors, errs := buildSelectorMap(ctx, c, objs)
+		if errs != nil {
+			return errs
+		}
+		var result []ast.FileObject
+		for _, obj := range objs.Namespace {
+			_, hasSelector := obj.GetAnnotations()[metadata.NamespaceSelectorAnnotationKey]
+			if hasSelector {
+				copies, err := makeNamespaceCopies(obj, nsSelectors)
+				if err != nil {
+					errs = status.Append(errs, err)
+				} else {
+					result = append(result, copies...)
+				}
 			} else {
-				result = append(result, copies...)
+				if obj.GetNamespace() == "" {
```

Comment:
`else if {`

Reply:
done

Reply:
reverted, because it appends the obj in the end o the else branch even if `obj.GetNamespace()` is not empty.